#' For RSEM data: read in and merge gene count data into a single matrix.
#'
#' Gene expressions are estimated by "abundance_estimates_to_matrix.pl" bundled with trinityrnaseq software using "--est_method RSEM".
#' @title RSEM data preparation: gene count data
#' @param d.path path to the folder containing "gene.count.matrix" files
#' @param o.path output file name.
#' @param g.type gene type: "gene" or "isoform"
#' @param d.type data type: "raw", "mean" or "sd". Out file of "raw" data can be applied directly in "run_DE_analysis.pl --method voom" downstream analysis.
#' @return matrix
#' @author zhao
#' @export
RSEM.gene_counts <- function(d.path, o.path=NULL, g.type="gene", d.type="raw"){
    patx <- paste0("\\.", g.type, "\\.counts\\.matrix$")
    xffs <- list.files(d.path, pattern=patx, full.names = TRUE)
    dtx <- NULL
    for (fx in xffs) {
        dxx <- read.table(fx, sep="\t", header=TRUE, row.names = 1)
        if(d.type == "mean") dxx <- rowMeans(dxx, na.rm = TRUE)
        else if(d.type == "sd") dxx <- unlist(apply(dxx, 1, sd, na.rm=TRUE))
        dtx <- if(is.null(dtx)) dxx else cbind(dtx, dxx)
    }
    if(d.type != "raw") colnames(dtx) <- sub(patx, "", basename(xffs))
    if(!is.null(o.path)) {
        if(d.type == "raw") write.table(dtx, o.path, sep="\t", quote = FALSE)
        else write.csv(dtx, o.path)
    }
    dtx
}

#' For RSEM-voom data: Get comparison data of with given treatment and control sample names.
#'
#' Raw data is generated by "run_DE_analysis.pl --method voom" (trinityrnaseq unitility)
#' @title RSEM data preparation: logFC and p values
#' @param d.path path to the folder containing "gene.count.matrix" files
#' @param comps.treatment sample names (treatments)
#' @param cmps.control sample names (controls)
#' @param o.path output file name.
#' @param g.type gene type: "gene" or "isoform"
#' @param d.type data type: "fc" for logFC and "p" for PValue column
#' @return data.frame whose first column is "gene_id"
#' @author zhao
#' @export
RSEM.comparison <- function(d.path, comps.treatment, cmps.control,
                            o.path=NULL, g.type="gene", d.type="fc"){
    spp <- cbind(comps.treatment, cmps.control)
    pref <- paste0("exprs.", g.type, ".")
    resval <- NULL
    cmps <- "gene_id"
    for(i in 1:nrow(spp)) {
        vs1 <- paste0(spp[i, 1], "_vs_", spp[i, 2])
        vs2 <- paste0(spp[i, 2], "_vs_", spp[i, 1])
        vss <- c(vs1, vs2)
        ffs <- file.path(d.path, paste0(pref, vss, ".voom.DE_results"))
        ffn <- file.exists(ffs)
        if(sum(ffn) < 1) next

        ffn <- which(ffn)[1]
        ## some comparisons may exist in reversed form.
        frev <- if(ffn == 1) 1 else -1
        cmps <- c(cmps, vs1) ## always use vs1 as label
        ## read data
        dtx <- read.table(ffs[ffn], header = TRUE)
        if(d.type == "p")
            vals <- data.frame(rownames(dtx), dtx$PValue)
        else
            vals <- data.frame(rownames(dtx), frev * dtx$logFC)
        colnames(vals) <- c("gene_id", paste0("V", i))
        if(is.null(resval)) resval <- vals
        else resval <- merge(resval, vals, by="gene_id", all = TRUE)
    }
    if(! is.null(resval)) colnames(resval) <- cmps
    if(!is.null(o.path)) write.csv(resval, o.path, row.names=FALSE)
    resval
}
